### PR TITLE
Issue 1398: Validated Controller URI at ClientConfig

### PR DESCRIPTION
**Pull request by a6dulaleem**
Signed-off-by: a6dulaleem <abdul.aleem1@dell.com>

**Change log description**  
Added URI validation in the clientConfig instead of nameResolver so that it can accept The URI which are not having any scheme like <host> : <port> and adding TCP as default scheme and throw IllegalArgumentException otherwise. 

**Purpose of the change**  
To fix Issue #1398 

**What the code does**  
Changed the clientConfig so that some Valid URI which are type of <HOST> : <PORT> are being accepted and no null pointer Exception would be generated. For this, Removed the Check from nameResolver and Added that check in clientConfig so that every ControllerURI must have scheme otherwise rejected for the other unsupported schemes.
This change is taking TCP as default scheme when no scheme is mentioned in ControllerURI for example : 
For URI localhost:9090 will be converted to  tcp://localhost:9090.

There might be some other URI which might be preceded by http, https, ftp etc in all those cases will throw IllegalArgumentException.

**How to verify it**  
The changes can be tested by running pravega standalone and providing the controllerURI as <localhost>:<9090> in the pravega properties or by giving in the AutoScallerConfig.java like this 

`this.controllerUri = URI.create("localhost:9090");``

In all above case the ControllerURI will be changed to `tcp://localhost:9090`

### CODE REVIEW

**Review by shshashwat — CHANGES_REQUESTED** [review]

**Comment by shshashwat** [comment]
@a6dulaleem , you need to add tests for this too

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 1]
```diff
@@ -230,6 +244,20 @@ public ClientConfig build() {
...
+                    !SCHEME_DIRECT.equals(scheme) && !SCHEME_DIRECT_SSL.equals(scheme)
+                    && !SCHEME_DIRECT_TLS.equals(scheme)) {
+                // If controllerURI is kind of <HOST>:<PORT> then will add TCP as default. 
+                // Otherwise URI might have some other Unsupported Scheme like http, https, ftp etc. in those cases will return null. 
+                String uri = controllerURI.toString();
+                if (uri.toString().split(":").length == 2) {
+                    controllerURI = URI.create("tcp://" + uri);
+                } else {
+                    log.warn("Unsupported URI: " + uri);
+                    return null;
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/ClientConfig.java:258]: This doesn't seem like the right way to handle this. It seems like this should throw IllegalArgumentException

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 2]
```diff
@@ -230,6 +244,20 @@ public ClientConfig build() {
             if (controllerURI == null) {
                 controllerURI = URI.create("tcp://localhost:9090");
             }
+            String scheme = controllerURI.getScheme();
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/ClientConfig.java:247]: It doesn't seem like the equals checks below are needed, if this properly returned null in the event of `host:port`.

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 3]
```diff
@@ -230,6 +244,20 @@ public ClientConfig build() {
...
+            if (!SCHEME_DISCOVER.equals(scheme) && !SCHEME_DISCOVER_TLS.equals(scheme) &&
+                    !SCHEME_DIRECT.equals(scheme) && !SCHEME_DIRECT_SSL.equals(scheme)
+                    && !SCHEME_DIRECT_TLS.equals(scheme)) {
+                // If controllerURI is kind of <HOST>:<PORT> then will add TCP as default. 
+                // Otherwise URI might have some other Unsupported Scheme like http, https, ftp etc. in those cases will return null. 
+                String uri = controllerURI.toString();
+                if (uri.toString().split(":").length == 2) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/ClientConfig.java:254]: Why is this parsing? URI already has functionality to get host and port.

**Review by tkaitchuck — CHANGES_REQUESTED** [review]

File: client/src/main/java/io/pravega/client/ClientConfig.java (hunk 2 above)
Comment (a6dulaleem) [reply client/src/main/java/io/pravega/client/ClientConfig.java:248]: resolved it.

File: client/src/main/java/io/pravega/client/ClientConfig.java (hunk 1 above)
Comment (a6dulaleem) [reply client/src/main/java/io/pravega/client/ClientConfig.java]: Added Illegal Argument Exception.

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 4]
```diff
@@ -230,6 +245,23 @@ public ClientConfig build() {
             if (controllerURI == null) {
                 controllerURI = URI.create("tcp://localhost:9090");
             }
+            String scheme = controllerURI.getScheme();
+            // If controllerURI is kind of <HOST>:<PORT> then will add TCP as default. 
+            // Otherwise if URI have some other Unsupported Scheme like http, https, ftp
+            // etc.
+            // in those cases will throw IllegalArgumentException.
```
Comment (shshashwat) [inline client/src/main/java/io/pravega/client/ClientConfig.java:252]: nit: a new line should start with caps

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 5]
```diff
@@ -230,6 +245,23 @@ public ClientConfig build() {
             if (controllerURI == null) {
                 controllerURI = URI.create("tcp://localhost:9090");
             }
+            String scheme = controllerURI.getScheme();
+            // If controllerURI is kind of <HOST>:<PORT> then will add TCP as default. 
+            // Otherwise if URI have some other Unsupported Scheme like http, https, ftp
+            // etc.
```
Comment (shshashwat) [inline client/src/main/java/io/pravega/client/ClientConfig.java:251]: nit: this line can be merged with the previous line

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 6]
```diff
@@ -230,6 +245,23 @@ public ClientConfig build() {
...
+            // in those cases will throw IllegalArgumentException.
+            if (!isValidScheme(scheme)) {
+                String uri = controllerURI.toString();
+                String[] uriParts = uri.split(":");
+                if (uriParts.length == 2) {
+                    controllerURI = URI.create(SCHEME_DIRECT + "://" + uriParts[0].replaceAll("/", "") + ":" + uriParts[1]);
+                    log.info("ControllerURI is kind of {} appending default scheme to it {}", uri,
+                            controllerURI.toString());
+                } else {
+                    throw new IllegalArgumentException(
```
Comment (shshashwat) [inline client/src/main/java/io/pravega/client/ClientConfig.java:261]: a `warn` level log can be put here too

**Review by shshashwat — CHANGES_REQUESTED** [review]

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 7]
```diff
@@ -230,6 +245,23 @@ public ClientConfig build() {
...
                 controllerURI = URI.create("tcp://localhost:9090");
             }
+            String scheme = controllerURI.getScheme();
+            // If controllerURI is kind of <HOST>:<PORT> then will add TCP as default. 
+            // Otherwise if URI have some other Unsupported Scheme like http, https, ftp
+            // etc.
+            // in those cases will throw IllegalArgumentException.
+            if (!isValidScheme(scheme)) {
+                String uri = controllerURI.toString();
+                String[] uriParts = uri.split(":");
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/ClientConfig.java:255]: This should be: 
`controllerURI.getHost()`
and later `controllerURI.getPort()`

**Review by shshashwat — APPROVED** [review]
LGTM

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 8]
```diff
@@ -230,6 +243,22 @@ public ClientConfig build() {
             if (controllerURI == null) {
                 controllerURI = URI.create("tcp://localhost:9090");
             }
+            String scheme = controllerURI.getScheme();
+            // If controllerURI is kind of <HOST>:<PORT> then will add TCP as default. 
+            // Otherwise if URI have some other Unsupported Scheme like http, https, ftp etc.
+            // In all those cases will throw IllegalArgumentException.
```
Comment (SrishT) [inline client/src/main/java/io/pravega/client/ClientConfig.java:249]: nit: Could this section be reworded?

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 9]
```diff
@@ -230,6 +243,22 @@ public ClientConfig build() {
...
+            if (!isValidScheme(scheme)) {
+                String uri = controllerURI.toString();
+                if (controllerURI.getScheme() != null && controllerURI.getHost() == null) {
+                    controllerURI = URI.create(SCHEME_DIRECT + "://" + uri);
+                    log.info("ControllerURI is kind of {} appending default scheme to it {}", uri,
```
Comment (SrishT) [inline client/src/main/java/io/pravega/client/ClientConfig.java:254]: nit: Phrase it as
```
("Scheme name is missing in the ControllerURI, therefore adding the default scheme {} to it", SCHEME_DIRECT)
```

**Review by tkaitchuck — APPROVED** [review]

**Pull request merged**